scripts/helpers.js

```diff
@@ -18,6 +19,52 @@ try {
 } catch (e) {}
 
 /**
+ * Generate data structure for generating table of contents.
+ *
+ * @param {string} content - Page content.
+ * @returns {array} [{title: 'Title', link: '#anchor', children: []}]
+ */
+hexo.extend.helper.register('table_of_contents', function (content) {
+  var $ = cheerio.load(content);
+  var h2Set = $('h2');
+  var toc = [];
+
+  // Get H2s.
+  h2Set.each(function (h2) {
+    var $h2 = $(this);
+
+    // Add H2.
+    var link = $h2.find('a')[0];
+    var item = {
+      title: link.attribs.title,
+      href: link.attribs.href,
+      children: []
+    };
+
+    // Get H3s. Loop until we hit an H2 or until no more siblings.
+    var next = $h2.next();
```

Comment:
can you use [`$h2.nextUntil(…)`](https://github.com/cheeriojs/cheerio#nextuntilselector-filter)?